packages/xstate-solid/src/utils.ts

```diff
@@ -0,0 +1,73 @@
+import type { AnyState, AnyStateMachine, State, StateFrom } from 'xstate';
+import type { SetStoreFunction } from 'solid-js/store';
+import { produce, reconcile } from 'solid-js/store';
+import rfdc from 'rfdc';
+
+// List of keys to reconcile while merging state
+const reconcileKeys: Array<keyof AnyState> = ['context', 'value'];
+
+/**
+ * Reconcile the state of the machine with the current state of the store.
+ * Handles primitive values, arrays, and objects.
+ * Provides granular reactivity for the state of the machine in solid-js.
+ */
+export const updateState = <
+  NextState extends StateFrom<AnyStateMachine> | object,
+  UpdateState extends SetStoreFunction<NextState>
+>(
+  nextState: NextState,
+  setState: UpdateState,
+  keys?: Array<keyof NextState>
+): void => {
+  if (typeof nextState === 'object' && !!nextState) {
+    // If a list of keys to update is not provided, get keys from nextState
+    if (!keys) {
+      keys = Object.keys(nextState) as Array<keyof NextState>;
+    }
+    for (const key of keys) {
+      // Don't update functions
+      if (typeof nextState[key] === 'function') {
+        continue;
+      }
+      if (key in nextState && reconcileKeys.includes(key as keyof State<any>)) {
```

Comment:
It worries me that this is done for all of the incoming updates. I think that, if possible, it would be best to only handle this particular type of logic when we are dealing with actual States. Perhaps this could be guarded based on the `instanceof` check (although I always try to avoid those) or just implemented in a dedicated function that would be used where we actually can receive States (like in `useMachine`)?

Reply:
Right now it's being used in `useMachine` both core and fsm, `useActor` and `useSelector`. We are already limiting it pretty heavily, only the 'context' property is being reconciled and everything else shallowly updated. We could limit it even further with a flag on the function so only State objects are reconciled. It would be more flexible and it could allow users to eventually opt in to reconciling on non State objects in the future.